File: examples/query-wrappers/QueryWrapper.py
from enum import Enum
from typing import Any, Callable, Optional, TypeVar, Generic

T = TypeVar('T')

class LogLevelType(Enum):
    INFO = 'INFO'
    WARN = 'WARN'
    ERROR = 'ERROR'
    FATAL = 'FATAL'

class CentralLogger:
    @staticmethod
    def log(level: LogLevelType, message: str, error: Optional[Exception] = None) -> None:
        print(f"[{level.value}] {message}: {error}")

class ErrorMessages(Enum):
    QUERY_FAILED = 'Database query failed'

class QueryResult(Generic[T]):
    def __init__(self, is_fail: bool, data: Optional[T] = None, error_message: Optional[str] = None):
        self.is_fail = is_fail
        self.data = data
        self.error_message = error_message

class QueryWrapper:
    @staticmethod
    def execute(query_fn: Callable[[], T]) -> QueryResult[T]:
        try:
            data = query_fn()
            return QueryResult(is_fail=False, data=data)
        except Exception as error:
            CentralLogger.log(LogLevelType.ERROR, ErrorMessages.QUERY_FAILED.value, error)
            return QueryResult(is_fail=True, data=None, error_message=ErrorMessages.QUERY_FAILED.value)

File: examples/query-wrappers/QueryWrapper.ts
enum LogLevelType {
  INFO = 'INFO',
  WARN = 'WARN',
  ERROR = 'ERROR',
  FATAL = 'FATAL'
}

class CentralLogger {
  static log(level: LogLevelType, message: string, error?: unknown): void {
    console.error(`[${level}] ${message}`, error);
  }
}

interface QueryResult<T> {
  data: T | null;
  isFail: boolean;
  errorMessage: string | null;
}

const ERROR_MESSAGES = {
  QUERY_FAILED: 'Database query failed'
} as const;

class QueryWrapper {
  static async execute<T>(queryFn: () => Promise<T>): Promise<QueryResult<T>> {
    try {
      const data = await queryFn();
      return {
        data,
        isFail: false,
        errorMessage: null
      };
    } catch (error) {
      CentralLogger.log(LogLevelType.ERROR, ERROR_MESSAGES.QUERY_FAILED, error);
      return {
        data: null,
        isFail: true,
        errorMessage: ERROR_MESSAGES.QUERY_FAILED
      };
    }
  }
}

export { QueryWrapper, LogLevelType, QueryResult, CentralLogger, ERROR_MESSAGES };
